flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala

```diff
@@ -449,14 +449,17 @@ object GenerateUtils {
    *   whether the input is nullable
    * @param deepCopy
    *   whether to copy the accessed field (usually needed when buffered)
+   * @param fusionCodegen
+   *   if fusion codegen enabled, don't need to hide generated code
```

Comment:
Why fusion codegen doesn't need to hide generated code? I mean why fusion codegen doesn't follow other's way to reuse the unboxing code and return expr with code erased?

Reply:
Currently, regarding the case of single operator codegen, we materialize the code for each input field to take values in advance, thus avoiding multiple evaluate them inside the operator. 
However, in the multi-operator fusion codegen case, the logic of our computation is reversed, and one of the objectives we want to achieve is to compute as lazy as possible, i.e., delay materialization. So we need to get the code that access the value of each field in advance, and then passes it to the downstream operator,  only materializing it where it is needed, instead of calculating it in advance. For example, for the Join operator, suppose there are 5 fields at the probe side and the build side together, and then pass to the downstream project operator, the project cuts out 3 of the fields and keeps only two, the three fields that are cut out, in fact, we do not need to take the value at all, which is one of the goals of OFCG, to eliminate invalid computation.

Reply:
I don't fully understand. If some fields can be pruned, why not the optimizer remove the fields?

Reply:
Giving an example, this is the plan
```
HashJoin(joinType=[InnerJoin], where=[=(a, d)], select=[a, d], build=[right])
:- Calc(select=[a])
:  +- HashJoin(joinType=[InnerJoin], where=[=(a, d)], select=[a, d], build=[right])
:     :- Exchange(distribution=[hash[a]])
:     :  +- Calc(select=[a])
:     :     +- BoundedStreamScan(table=[[default_catalog, default_database, x]], fields=[a, b, c, nx])
:     +- Exchange(distribution=[hash[d]])
:        +- Calc(select=[d])
:           +- BoundedStreamScan(table=[[default_catalog, default_database, y]], fields=[d, e, f, ny])
+- Calc(select=[d])
   +- HashJoin(joinType=[InnerJoin], where=[=(d, a)], select=[d, a], build=[right])
      :- Exchange(distribution=[hash[d]])
      :  +- Calc(select=[d])
      :     +- BoundedStreamScan(table=[[default_catalog, default_database, y]], fields=[d, e, f, ny])
      +- Exchange(distribution=[hash[a]])
         +- Calc(select=[a])
            +- BoundedStreamScan(table=[[default_catalog, default_database, t]], fields=[a, b, c, nt])

```
In the Join case, field `a` from the join probe side, and field `d` from the build side,  but the downstream join operator only needs to use field `a`, the `d` is unnecessary. However, the join operator can't prune the `d` itself, so the optimizer inserts a calc operator between two join operators to prune the field `d`. In this case, evaluating field `d` is useless, so we can skip it.